### PR TITLE
feat: handle promise based functions

### PR DESCRIPTION
**Pull request by AVaksman**
Paginator will now handle `promise` based functions.

It is a prerequisite for converting libraries to `async` first approach.

- [x] Tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)

### CODE REVIEW

**Comment by AVaksman** [comment]
To keep unrelated changes separate, the failing docs test is addressed in PR #92. After #92 lands will merge master here.

File: src/index.ts [hunk 1]
```diff
@@ -265,12 +265,21 @@ export class Paginator {
     const callback = parsedArguments.callback!;
     if (parsedArguments.autoPaginate) {
       const results = new Array<{}>();
-      paginator.runAsStream_(parsedArguments, originalMethod)
-          .on('error', callback)
-          .on('data', (data: {}) => results.push(data))
-          .on('end', () => callback(null, results));
+      if (callback) {
```
Comment (callmehiphop) [inline src/index.ts]: Can we just make one call to `runAsStream_` wrapped in a promise and if we detect a callback do something like 

```js
promise.then(results => callback(null, ...results), callback);
```

Comment (AVaksman) [reply src/index.ts]: Thanks, changed the code accordingly. PTAL.

**Comment by JustinBeckwith** [comment]
@callmehiphop can you give this another look and merge if happy?

File: src/index.ts [hunk 2]
```diff
@@ -265,12 +265,21 @@ export class Paginator {
...
+        paginator.runAsStream_(parsedArguments, originalMethod)
+            .on('error', reject)
+            .on('data', (data: {}) => results.push(data))
+            .on('end', () => resolve([results]));
+      });
+      if (callback) {
+        promise.then(
+            results => callback(null, ...results as []),
+            (err: Error) => callback(err));
+      } else {
```
Comment (callmehiphop) [inline src/index.ts]: This is purely a nitpick, but all the `if/else` action going on here feels unnecessary. Could we simplify the code by doing something like this

```typescript
if (!parsedArguments.autoPaginate) {
  return originalMethod(query, callback);
}

const promise = new Promise((resolve, reject) => {
  // ...
});

if (!callback) {
  return promise;
}

promise.then(/*...*/);
```

File: src/index.ts [hunk 3]
```diff
@@ -308,23 +317,29 @@ export class Paginator {
     });
 
     function makeRequest(query?: ParsedArguments|string) {
-      originalMethod(query, onResultSet);
```
Comment (callmehiphop) [inline src/index.ts]: Is this change actually needed? Unless we have plans to drop callback support entirely, I think it is probably safe to leave this alone, no?

File: src/index.ts [hunk 4]
```diff
@@ -265,12 +265,21 @@ export class Paginator {
...
       const results = new Array<{}>();
-      paginator.runAsStream_(parsedArguments, originalMethod)
-          .on('error', callback)
-          .on('data', (data: {}) => results.push(data))
-          .on('end', () => callback(null, results));
+      const promise = new Promise((resolve, reject) => {
+        paginator.runAsStream_(parsedArguments, originalMethod)
+            .on('error', reject)
+            .on('data', (data: {}) => results.push(data))
+            .on('end', () => resolve([results]));
```
Comment (callmehiphop) [inline src/index.ts]: Another nit I have is that we're packaging this up to match the expected return value of a Promise before we even know what the user wants. I think it would be slightly more efficient if we just packed is up via promise chain

```js
return promise.then(results => [results]);
```

File: src/index.ts [hunk 5]
```diff
@@ -221,8 +221,8 @@ export class Paginator {
         delete query.maxApiCalls;
       }
 
-      // maxResults is the user specified a limit.
-      if (callback && (maxResults !== -1 || query.autoPaginate === false)) {
```
Comment (callmehiphop) [inline src/index.ts]: I've been thinking about this change a bit and I have some concerns. Since we remove the callback check I think this means that this logic will now apply to any object streams. Hopefully @stephenplusplus can chime in here but I think that this means setting `maxResults` on an object stream method will cause it to come in under the desired amount since now we'll be flagging `autoPaginate` to false.

Comment (callmehiphop) [reply src/index.ts]: For a follow up thought, since callback was previously used to determine if we wanted a stream or not, maybe we should create an internal `streaming` flag that we check in place of this. 

```typescript
parseArguments: (_args: any[], streaming: boolean);
```

File: src/index.ts (hunk 3 above)
Comment (AVaksman) [reply src/index.ts]: I made this change in order for `paginator` not to be dependent on `callbackify` and be able to handle a truly pure promise base function.
Also in hopes, that it will be more efficient this way rather than funnel every function through `callbackify` every time.

Let me know your preference though.

File: src/index.ts (hunk 4 above)
Comment (AVaksman) [reply src/index.ts]: done, thanks

File: src/index.ts (hunk 2 above)
Comment (AVaksman) [reply src/index.ts]: done, thanks

File: src/index.ts (hunk 3 above)
Comment (callmehiphop) [reply src/index.ts]: Won't `parsedArguments.callback` be the user provided callback function though? Using the user function to determine which version (promise/callback) to call means that you will definitely be dependent on callbackify for promise first clients. I think if your goal is to not be dependent on callbackify/promisify you should probably always use the promise version under the assumption that callback first will eventually go away.

File: src/index.ts (hunk 5 above)
Comment (AVaksman) [reply src/index.ts]: `autoPaginate` is only used by `paginator.extend` to determine whether we wanted a stream or not. 
`autoPaginate` could get flagged to `false` only by user setting `maxResults`/`pageSize` to a value or setting `autoPaginate` to `false`. Previously all methods came with `callback`s (even if user did not provide one). Now since we are introducing promise based methods, `callback` becomes optional, and we might run in an issue of not flagging `autoPaginate` to false when it is needed.

Comment (callmehiphop) [reply src/index.ts]: > autoPaginate is only used by paginator.extend to determine whether we wanted a stream or not.

This isn't quite accurate,`autoPaginate` is used to determine if we should make multiple requests. A separate method (`Paginator.streamify()`) is used for generating streams, which also calls `parseArguments_`.

> Previously all methods came with callbacks (even if user did not provide one). 

This is true for callback/promise methods, but `callback` has always been undefined when called from `Paginator.streamify()`.

> Now since we are introducing promise based methods, callback becomes optional, and we might run in an issue of not flagging autoPaginate to false when it is needed.

That's why I suggested adding a second boolean parameter to determine if the user wants a stream or not. Setting said parameter to `false` inside of `Paginator.extend()` and `true` inside of `Paginator.streamify()`.

Comment (callmehiphop) [reply src/index.ts]: Actually, I just realized `runAsStream_` ignores `autoPaginate` so I guess it doesn't matter. Feel free to ignore this and sorry about the confusion!

File: src/index.ts (hunk 3 above)
Comment (AVaksman) [reply src/index.ts]: I guess what I wanted to say is, in case when user didn't provide callback in the original method, I didn't want `paginator` to provide one. This way `callbackify` will just execute original method as is, instead of going through the process of removing imposed `cb` from `arguments`, still executing as promise and pass the `arrified` response to callback.
And also in remote situation when someone will use `paginator` to extend just a promise (not callbackified) it wouldn't crush.

Comment (callmehiphop) [reply src/index.ts]: Using the streaming APIs as an example, `parsedArguments.callback` will always be`undefined`, right? This logic then assumes that since there is no callback, the original method must be async first, yes?

Comment (AVaksman) [reply src/index.ts]: Got it, done
thanks

Comment (callmehiphop) [reply src/index.ts]: Oh man, sorry if I've been a pain in the butt! I liked the optimization, I would have just preferred an option or something to indicate which version to use. Maybe we should open an issue that says to convert this to promise only after all the clients have been updated.
